app/src/org/commcare/dalvik/activities/PinAuthenticationActivity.java

```diff
@@ -0,0 +1,173 @@
+package org.commcare.dalvik.activities;
+
+import android.content.Intent;
+import android.os.Bundle;
+import android.util.Log;
+import android.view.View;
+import android.widget.Button;
+import android.widget.EditText;
+import android.widget.TextView;
+import android.widget.Toast;
+
+import org.commcare.android.database.app.models.UserKeyRecord;
+import org.commcare.android.framework.ManagedUi;
+import org.commcare.android.framework.SessionAwareCommCareActivity;
+import org.commcare.android.framework.UiElement;
+import org.commcare.dalvik.R;
+import org.commcare.dalvik.application.CommCareApplication;
+import org.javarosa.core.services.locale.Localization;
+
+/**
+ * When a user selects Set/Reset PIN from the home activity options menu, they must first
```

Comment:
@ctsims This is actually necessary. In general, whenever we want to assign a new PIN to a user, we need to have access to the un-hashed password in order to do so. When the user selects Set/Reset PIN from the options menu, we are at a point where we no longer have access to that un-hashed password (since they could go into the options menu at any point, well after the initial login), and therefore need to require them to enter either their PW or current PIN, in order to get it. 

The use case you're describing of having just logged in and then needing to re-enter your password shouldn't really come up because if you've logged in in password mode, and the PIN setting is enabled, the pin choice dialog will show up immediately upon login, and if you select "Yes, create a pin" from THAT screen, you don't have to re-auth. You only get shown the PinAuthenticationActivity if you select Set/Reset PIN from the home activity options menu.